base/system/winlogon/sound.c

```diff
@@ -0,0 +1,325 @@
+
+    if (hLibrary)
+    {
+        _waveOutGetNumDevs = (PFWAVEOUTGETNUMDEVS)GetProcAddress(hLibrary, "waveOutGetNumDevs");
+        
+        if (_waveOutGetNumDevs)
+        {
+            uNumDevs = _waveOutGetNumDevs();
+
+            if (uNumDevs == 0)
+            {
+                if (!bLogon)
+                {
+                    Beep(440, 125);
+                }
+
+                FreeLibrary(hLibrary);
+
+                return FALSE;
+            }
+        }
+
+        _PlaySoundW = (PFPLAYSOUNDW)GetProcAddress(hLibrary, "PlaySoundW");
+
+        if (_PlaySoundW)
+        {
+            bRet = _PlaySoundW(lpFileName, NULL, dwFlags);
+        }
+
+        FreeLibrary(hLibrary);
+    }
+
+    return bRet;
+}
+
+static
+DWORD
+WINAPI
+PlaySystemSoundThread(
+    IN LPVOID lpParameter)
+{
+    PWL_PLAYSOUND_DATA PSData;
+    BOOL bLogon;
+    LPWSTR pszSndKey;
+    HKEY hUserKey, hSndKey;
+    DWORD dwType, dwSize;
+    LONG lError;
+    LPWSTR pszValue = NULL;
+    LPWSTR pszSndPath = NULL;
+
+    PSData = (PWL_PLAYSOUND_DATA)lpParameter;
+    bLogon = (PSData->Sound == SYSTEMSND_LOGON);
+
+    if (bLogon)
+    {
+        SERVICE_STATUS_PROCESS Info;
+        DWORD dwBytesNeeded;
+        ULONG Index = 0;
+        SC_HANDLE hSCManager, hService;
+
+        /* Open the service manager */
+        hSCManager = OpenSCManagerW(NULL, NULL, SC_MANAGER_CONNECT);
+
+        if (!hSCManager)
+        {
+            ERR("OpenSCManagerW failed (%x)\n", GetLastError());
+            goto Cleanup;
+        }
+
+        /* Open the wdmaud service */
+        hService = OpenServiceW(hSCManager, L"wdmaud", GENERIC_READ);
+
+        if (!hService)
+        {
+            /* The service is not installed */
+            TRACE("Failed to open wdmaud service (%x)\n", GetLastError());
+            CloseServiceHandle(hSCManager);
+            goto Cleanup;
+        }
+
+        /* Wait for wdmaud to start */
+        do
+        {
+            if (!QueryServiceStatusEx(hService, SC_STATUS_PROCESS_INFO, (LPBYTE)&Info, sizeof(Info), &dwBytesNeeded))
+            {
+                TRACE("QueryServiceStatusEx failed (%x)\n", GetLastError());
+                break;
+            }
+
+            if (Info.dwCurrentState == SERVICE_RUNNING)
+            {
+                break;
+            }
+
+            Sleep(1000);
+
+        } while (Index++ < 20);
+
+        CloseServiceHandle(hService);
+        CloseServiceHandle(hSCManager);
+
+        /* If wdmaud is not running exit */
+        if (Info.dwCurrentState != SERVICE_RUNNING)
+        {
+            WARN("wdmaud has not started!\n");
+            goto Cleanup;
+        }
+
+        /* Sound subsystem is running. Play logon sound. */
+    }
+
+    /* Impersonate current user */
+    if (!ImpersonateLoggedOnUser(PSData->Session->UserToken))
+    {
+        goto Cleanup;
+    }
+
+    /* Open user's HKCU */
+    lError = RegOpenCurrentUser(KEY_READ, &hUserKey);
+
+    RevertToSelf();
```

Comment:
Should I move this `RevertToSelf` to `Cleanup` section?

Reply:
Why would you do that?

Reply:
For readability, maybe.